infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/receive/StreamsSnapshotWriter.java

```diff
@@ -0,0 +1,273 @@
+        }
+
+        byte[] payload = message.getPayload();
+        OpaqueEntry opaqueEntry = OpaqueEntry.deserialize(Unpooled.wrappedBuffer(payload));
+
+        if (opaqueEntry.getEntries().keySet().size() != 1) {
+            log.error("The opaqueEntry has more than one entry {}", opaqueEntry);
+            return;
+        }
+
+        UUID uuid = opaqueEntry.getEntries().keySet().stream().findFirst().get();
+        processOpaqueEntry(opaqueEntry.getEntries().get(uuid), message.getMetadata().getSnapshotSyncSeqNum(), uuidMap.get(uuid));
+        recvSeq++;
+    }
+
+    @Override
+    public void apply(List<LogReplicationEntry> messages) throws Exception {
+        for (LogReplicationEntry msg : messages) {
+            apply(msg);
+        }
+    }
+
+    /**
+     * Read from the shadow table and write to the real table
+     * @param uuid: the real table uuid
+     */
+    public long applyShadowStream(UUID uuid, Long seqNum, long snapshot) {
+        UUID shadowUUID = uuidMap.get(uuid);
+        StreamOptions options = StreamOptions.builder()
+                .ignoreTrimmed(false)
+                .cacheEntries(false)
+                .build();
+
+        //Can we do a seek after open to ignore all entries that are earlier
+        Stream shadowStream = (new OpaqueStream(rt, rt.getStreamsView().get(shadowUUID, options))).streamUpTo(snapshot);
+
+        Iterator<OpaqueEntry> iterator = shadowStream.iterator();
+        while (iterator.hasNext()) {
+            OpaqueEntry opaqueEntry = iterator.next();
+            if (opaqueEntry.getVersion() > shadowStreamStartAddress) {
+                processOpaqueEntry(opaqueEntry.getEntries().get(shadowUUID), seqNum, uuid);
+                seqNum = seqNum + 1;
+            }
+        }
+
+        return seqNum;
+    }
+
+
+    /**
+     * read from shadowStream and append to the
+     */
+    public void applyShadowStreams(Long seqNum) {
+        phase = Phase.ApplyPhase;
+        long snapshot = rt.getAddressSpaceView().getLogTail();
+        clearTables();
+        for (UUID uuid : streamViewMap.keySet()) {
+            seqNum = applyShadowStream(uuid, seqNum, snapshot);
+        }
+    }
+
+    /**
+     * Snapshot data has been transferred from primary node to the standby node
+     * @param entry
+     */
+    public  void snapshotTransferDone(LogReplicationEntry entry) {
+        phase = Phase.ApplyPhase;
+        //verify that the snapshot Apply hasn't started yet and set it as started and set the seqNumber
+        long ts = entry.getMetadata().getSnapshotTimestamp();
+        long seqNum = 0;
+        siteConfigID = entry.getMetadata().getSiteConfigID();
+
+        //update the metadata
+        persistedWriterMetadata.setLastSnapTransferDoneTimestamp(siteConfigID, ts);
+
+        //get the number of entries to apply
+        seqNum = 1 + persistedWriterMetadata.query(null, PersistedWriterMetadata.PersistedWriterMetadataType.LastSnapSeqNum);
```

Comment:
what is this seqNum , add comments. Looks like the snapshotSeqNum to identify this snapshotSync.